tests/js/server/aql/aql-queries-optimizer-count.js

```diff
@@ -0,0 +1,105 @@
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Heiko Kernbach
+/// @author Copyright 2021, ArangoDB GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+
+let jsunity = require("jsunity");
+let internal = require("internal");
+let helper = require("@arangodb/aql-helper");
+let getQueryResults = helper.getQueryResults;
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief test suite
+////////////////////////////////////////////////////////////////////////////////
+
+function ahuacatlQueryOptimizerCollectTestSuite() {
+  let collection = null;
+  let cn = "UnitTestsAhuacatlOptimizerCountCollect";
+
+  return {
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief set up
+////////////////////////////////////////////////////////////////////////////////
+
+    setUp: function () {
+      internal.db._drop(cn);
+      collection = internal.db._create(cn);
+
+      // create an index with two attributes (first: boolean, second: array)
+      collection.ensureIndex({type: "persistent", fields: ["someBoolean", "someArray[*]"]});
+
+      // create 100 (50+50) documents in total
+      for (let outer = 0; outer < 50; outer++) {
+        let myArray = [];
+        // create 100 entries for our inner document array
+        for (let inner = 0; inner < 10; inner++) {
+          // write numeric strings
+          myArray.push(JSON.stringify(inner));
+        }
+        collection.save({"someBoolean": true, "someArray": myArray});
+        collection.save({"someBoolean": false, "someArray": myArray});
+      }
+    },
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief tear down
+////////////////////////////////////////////////////////////////////////////////
+
+    tearDown: function () {
+      internal.db._drop(cn);
+    },
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief check sort optimization without index
+////////////////////////////////////////////////////////////////////////////////
+
+    testCountWithMultipleAttributesIncludingIndexArray: function () {
+      let countWithLengthQuery = "RETURN LENGTH(" + cn + ")";
+      let countWithCollectQuery = "FOR d IN " + cn + " FILTER (d.someBoolean == true) OR (d.someBoolean == false) COLLECT WITH COUNT INTO c RETURN c";
+
+      let resCountWithLength = getQueryResults(countWithLengthQuery);
+      let resCountWithCollect = getQueryResults(countWithCollectQuery);
+
+      assertEqual(1, resCountWithLength.length);
+      assertEqual(1, resCountWithCollect.length);
+      assertEqual(resCountWithLength[0], 100);
+      assertEqual(resCountWithCollect[0], 100);
+
+      assertEqual(resCountWithCollect[0], resCountWithLength[0]);
+    }
+  };
```

Comment:
There should be an assertion that the index is actually used by the `FOR` operation to make sure we test the correct thing. Or alternatively, add an [index hint](https://www.arangodb.com/docs/stable/aql/operations-for.html#indexhint) with `forceIndexHint: true` for the same effect.